ddl/partition.go

```diff
@@ -2137,6 +2164,589 @@ func (w *worker) onExchangeTablePartition(d *ddlCtx, t *meta.Meta, job *model.Jo
+}
+
+func (w *reorgPartitionWorker) cleanRowMap() {
+	for id := range w.rowMap {
+		delete(w.rowMap, id)
+	}
+}
+
+func (w *reorgPartitionWorker) AddMetricInfo(cnt float64) {
+	w.metricCounter.Add(cnt)
+}
+
+// updateCurrentElement update the current element for reorgInfo.
+func (w *worker) reorgPartitionData(t table.Table, reorgInfo *reorgInfo) error {
+	// TableIDs should already been assigned (todo: remove this note)
+	// First copy all table data to the new partitions
+	// from each of the DroppingDefinitions partitions
+	// Create all indexes on the AddingDefinitions partitions
+
+	/*
+			failpoint.Inject("mockInfiniteReorgLogic", func(val failpoint.Value) {
+			//nolint:forcetypeassert
+			if val.(bool) {
+				a := new(interface{})
+				TestReorgGoroutineRunning <- a
+				for {
+					time.Sleep(30 * time.Millisecond)
+					if w.getReorgCtx(reorgInfo.Job).isReorgCanceled() {
+						// Job is cancelled. So it can't be done.
+						failpoint.Return(dbterror.ErrCancelledDDLJob)
+					}
+				}
+			}
+		})
+	*/
+
+	// Copy the data from the DroppingDefinitions to the AddingDefinitions
+	if bytes.Equal(reorgInfo.currElement.TypeKey, meta.ColumnElementKey) {
+		err := w.updatePhysicalTableRow(t, reorgInfo)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	// Rewrite this to do all indexes at once in addTableIndex
+	// instead of calling it once per index (meaning reading the table multiple times)
+	// But for now, try to understand how it works...
+	firstNewPartitionID := t.Meta().Partition.AddingDefinitions[0].ID
+	startElementOffset := 0
+	//startElementOffsetToResetHandle := -1
+	// This backfill job starts with backfilling index data, whose index ID is currElement.ID.
+	if !bytes.Equal(reorgInfo.currElement.TypeKey, meta.IndexElementKey) {
+		// First run, have not yet started backfilling index data
+		// Restart with the first new partition.
+		// TODO: handle remove partitioning
+		reorgInfo.PhysicalTableID = firstNewPartitionID
+	} else {
+		// The job was interrupted and has been restarted,
+		// reset and start from where it was done
+		for i, element := range reorgInfo.elements[1:] {
+			if reorgInfo.currElement.ID == element.ID {
+				startElementOffset = i
+				//startElementOffsetToResetHandle = i
+				break
+			}
+		}
+	}
+
+	for i := startElementOffset; i < len(reorgInfo.elements[1:]); i++ {
+		// Now build the indexes in the new partitions
+		var physTbl table.PhysicalTable
+		if tbl, ok := t.(table.PartitionedTable); ok {
+			physTbl = tbl.GetPartition(reorgInfo.PhysicalTableID)
+		} else if tbl, ok := t.(table.PhysicalTable); ok {
+			// This may be used when partitioning a non-partitioned table
+			physTbl = tbl
+		}
+		// Get the original start handle and end handle.
+		currentVer, err := getValidCurrentVersion(reorgInfo.d.store)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		// TODO: Can we improve this incase or a crash?
```

Comment:
I can see a lot of questions like this in this PR... what is your plan to get them resolved? Will them be solved after merging this(and the succeeding PRs) into the feature branch?

Reply:
Hi, my plan is to work on this feature in four steps (any suggestion on how to explain it better in https://github.com/pingcap/tidb/issues/38535 is welcome):
1) Implement a working reorganize data and index DDL backfiller (this PR) Only focusing on the actual data copying and the index regeneration.
2) Complete the concurrent write traffic / double write for foreground traffic during the stateWriteOnly/stateWriteReorganization states (pr 2, https://github.com/pingcap/tidb/pull/38508 )
3) Finalise the reorg framework (how the different states and various rollback scenarios works) (pr 3, wip, https://github.com/pingcap/tidb/pull/38532 )
4) Wrap everything up and add missing tests and remove loose ends (like all the added TODOs) coming PR.